### PR TITLE
refactor: use `SourceLocation` in `QName`

### PR DESCRIPTION
**Pull request by magnus-madsen**
Towards only using source locations in Name

### CODE REVIEW

**Comment by magnus-madsen** [comment]
A careful review to see that I did not screw up would be useful.

File: main/src/ca/uwaterloo/flix/language/ast/Name.scala [hunk 1]
```diff
@@ -26,31 +26,25 @@ object Name {
...
     * Creates a qualified name from the given namespace `ns` and name `name`.
     */
-  def mkQName(ns: List[String], name: String, sp1: SourcePosition, sp2: SourcePosition): QName = {
-    val loc = SourceLocation(isReal = true, sp1, sp2)
-    val nname = NName(ns.map(t => Name.Ident(sp1, t, sp2)), loc)
-    val ident = Ident(sp1, name, sp2)
-    QName(sp1, nname, ident, sp2)
+  def mkQName(ns: List[String], name: String, loc: SourceLocation): QName = {
+    val nname = NName(ns.map(t => Name.Ident(loc.sp1, t, loc.sp2)), loc)
+    val ident = Ident(loc.sp1, name, loc.sp2)
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/Name.scala:44]: Will we change Ident to use `SourceLocation` too?

File: main/src/ca/uwaterloo/flix/language/phase/Desugar.scala [hunk 2]
```diff
@@ -1135,7 +1135,7 @@ object Desugar {
     // ignore redundant effect casts to IO
     val correctedEff = eff match {
       case None => None
-      case Some(DesugaredAst.Type.Ambiguous(Name.QName(_, Name.NName(Nil, _), Name.Ident(_, "IO", _), _), _)) => None
+      case Some(DesugaredAst.Type.Ambiguous(Name.QName(Name.NName(Nil, _), Name.Ident(_, "IO", _), _), _)) => None
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Desugar.scala:1138]: Can we just check that the name is unqualified and matches IO? This is hard to read.

File: main/src/ca/uwaterloo/flix/language/ast/Name.scala [hunk 3]
```diff
@@ -26,31 +26,25 @@ object Name {
...
+    QName(nname, ident, loc)
   }
 
-  /**
-    * Returns the given identifier `ident` as qualified name in the root namespace.
-    */
-  def mkQName(ident: Ident): QName = QName(ident.sp1, RootNS, ident, ident.sp2)
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/Name.scala:52]: Why did we get rid of this helper? Seems like it's used a lot.

**Review by mlutze — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/ast/Name.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Name.scala:44]: Yes-- thats my ultimate plan. I am taking small steps towards that.

File: main/src/ca/uwaterloo/flix/language/phase/Desugar.scala (hunk 2 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Desugar.scala:1138]: I suppose I can refactor.

File: main/src/ca/uwaterloo/flix/language/ast/Name.scala (hunk 3 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Name.scala:52]: My plan is to get rid of these smart constructors and focus _on the data_.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Name.scala:52]: We could an expression to deal with that.

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Name.scala:52]: IMO the root of the problem is that our standard library is unqualified. We should just have a `Flix` namespace and put the who stdlib in it.

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Name.scala:52]: Then we can fully separate qualified and unqualified names

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Name.scala:52]: This is orthogonal though

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Name.scala:52]: Ok maybe that's actually orthogonal, but we should do it anyway :)

**Pull request merged**